File: disciple/win32/win32_ipc.ml
(* OCaml-Win32
 * win32_ipc.ml
 * Copyright (c) 2002 by Harry Chomsky
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *)

open Win32_core

(********************************************************************)
(* User Interface Services: User Input: Keyboard Accelerators *)

(* Messages *)

let wm_copydata = 0x004A
type wm_copydata_handler =
    wnd:hwnd ->
    msg:message_contents ->
    source_wnd:hwnd ->
    cds:copydatastruct ->
    bool message_handler_result
external deliver_wm_copydata :
    handler:wm_copydata_handler ->
    wnd:hwnd ->
    msg:message_contents ->
    bool message_handler_result = "deliver_wm_copydata"
external return_wm_copydata : result:bool -> message_result = "return_wm_copydata"
let on_wm_copydata = on_message ~id:wm_copydata ~deliver:deliver_wm_copydata ~return:return_wm_copydata
external issue_wm_copydata : sender:message_sender -> source_wnd:hwnd -> cds:copydatastruct -> bool = "issue_wm_copydata"

File: disciple/win32/win32_ipc.c
/* OCaml-Win32
 * win32_ipc.c
 * Copyright (c) 2002 by Harry Chomsky
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "win32.h"

/********************************************************************/
/* User Interface Services: User Input: Keyboard Accelerators */

/* Messages */

CAMLprim value deliver_wm_copydata(value handler, value wnd, value msg)
{
    CAMLparam3(handler, wnd, msg);
    CAMLlocalN(args, 4);
    CAMLlocal1(result);

    args[0] = wnd;
    args[1] = msg;
    args[2] = alloc_handle((HWND)Wparam(msg));
    args[3] = alloc_copydatastruct((PCOPYDATASTRUCT)Lparam(msg));
    result = callbackN(handler, 4, args);
    CAMLreturn(result);
}

CAMLprim value return_wm_copydata(value result)
{
    return copy_int32(Bool_val(result));
}

CAMLprim value issue_wm_copydata(value sender, value source_wnd, value cds)
{
    CAMLparam3(sender, source_wnd, cds);
    COPYDATASTRUCT c;
    CAMLlocal2(contents, result);

    fill_copydatastruct(&c, cds);
    contents = pack_message(WM_COPYDATA, (WPARAM)Handle_val(source_wnd), (LPARAM)&c);
    result = callback(sender, contents);
    empty_copydatastruct(&c, cds);
    CAMLreturn(Val_bool(Int32_val(result)));
}
